### PR TITLE
Temporarily avoid a panic when downcasting to redjubjub::Error fails

### LINKED ISSUE #1357: Panic on failed downcast from BoxError to redjubjub::Error
**Version**

`zebrad` main branch, 2020-11-23.

**Platform**

`Linux ... 5.4.75 #1-NixOS SMP Thu Nov 5 10:43:38 UTC 2020 x86_64 GNU/Linux`

**Description**

When syncing Mainnet with `checkpoint_sync = true` around block 1018849:
1. Zebra panics due to a failed downcast from `BoxError` to `redjubjub::Error`
2. Using `panic!(e)` displays an unhelpful `<non string panic payload>` panic message

Note: 1018849 is the second-last checkpoint when `checkpoint_sync = true`.

The failing code is:
```rust
impl From<BoxError> for TransactionError {
    fn from(err: BoxError) -> Self {
        match err.downcast::<redjubjub::Error>() {
            Ok(e) => TransactionError::RedJubjub(*e),
            Err(e) => panic!(e),
        }
    }
}
```

The logs are:
```
Nov 23 15:38:42.088  INFO sync:extend_tips:download_and_verify{hash=block::Hash("00000000021615cd8f0425dfa64960804f68eeba533270c02bb44e8b9bea2420")}:checkpoint: zebra_consensus::checkpoint: verified checkpoint range block_count=400 current_range=(Excluded(Height(1018049)), Included(Height(1018449)))                                                                                  
Nov 23 15:38:46.022  INFO sync: zebrad::components::sync: waiting for pending blocks tips.len=1 in_flight=2372 lookahead_limit=2000                                                            
Nov 23 15:38:46.023  INFO sync: zebrad::components::sync: extending tips tips.len=1 in_flight=2000 lookahead_limit=2000                                                                        
Nov 23 15:38:46.479  INFO sync:extend_tips:download_and_verify{hash=block::Hash("00000000005b1fd9b021a22f493818246efd17b77a50a524e17b570c1d55bbfa")}:checkpoint: zebra_consensus::checkpoint: verified checkpoint range block_count=400 current_range=(Excluded(Height(1018449)), Included(Height(1018849)))                                                                                  
The application panicked (crashed).                                                                                                                                                            
Message:  <non string panic payload>
Location: zebra-consensus/src/error.rs:76
  
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ SPANTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
      
   0: zebrad::components::sync::downloads::download_and_verify with hash=block::Hash("00000000002c76519f671c6789f99af6345fac144509863fb01185cab68198f9")
      at zebrad/src/components/sync/downloads.rs:113
   1: zebrad::components::sync::extend_tips
      at zebrad/src/components/sync.rs:381
   2: zebrad::components::sync::sync
      at zebrad/src/components/sync.rs:150

```

**Related Issues**

Closed by #1363.
Cleanup underlying cause in #1186

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

Zebra panics with a downcast error due to bug #1357.

We want to do a permanent fix as part of #1186, but that change will take more time, and it is not as urgent.

Right now, the panic is blocking other testing.

## Solution

Turn the panic into a temporary internal error type, which contains the original error formatted as a string. This downgrades the panic to a log message, with a level chosen by the code that processes the error. (Typically info-level.)

The code in this pull request has:
  - [x] Documentation Comments

## Review

@hdevalence is working on the permanent fix.

I need this panic fixed urgently, so I can complete a full sync to test the difficulty contextual validation. I expect that others also want these panics fixed, so they can do tests.

## Related Issues

Closes #1357
Cleanup underlying issue in #1186

## Follow Up Work

The temporary internal error type should be removed when #1186 is fixed.

### CODE REVIEW

**Review by hdevalence — APPROVED** [review]
LGTM

**Comment by teor2345** [comment]
(I'm just going to wait for CI to pass before I merge.)

**Review by dconnolly — APPROVED** [review]

**Pull request merged**
